yappi/_yappi.c

```diff
@@ -1263,6 +1366,10 @@ profile_event(PyObject *self, PyObject *args)
         return NULL;
     }
 
+    // TODO(suhail): Understand how it is possible to reach here when
+    // 'flags.multithreaded' is false. Also if multithreaded is false, then
+    // this is definitely not the initial thread that called 'start' and so
+    // we should not continue to invoke '_yapp_callback' as we do now.
     if (flags.multithreaded) {

```

Comment:
But this only checks for `multithreaded==true`. False is not possible as you indicate. That if check is there just for extra readability/verbosity? Not sure :) . THis is for the case, where we have started yappi and some other thread started running and a profile event is called for the new thread. _ensure_thread_profiled() is always valid for multithreaded profiling.

Reply:
Oh alright. I assumed it was added there to handle an edge case when multithreaded is false but I could neither understand how it would be possible nor how this would fix anything if hit. I'll remove it in case somebody else stumbles upon the same problem.

Reply:
Re-thinking about the problem again I think the correct behaviour should be:

```
if (!flags.multithreaded) {
    Py_RETURN_NONE;
}
```

LIke you indicated, right? I was mistaken there is a way this code might be called with multiple threads. However, inside `_current_context_id` there is another check for this which fixes the problem:

```
if (!flags.multithreaded) {
     return 0;
}
```

This makes sure only single thread is profiled. But like you suggested, it would be better if we check for this condition early in the `profile_event` function.